qiling/os/uefi/uefi.py

```diff
@@ -0,0 +1,242 @@
+                    self.ql.entry_point = entry_point
+                self.ql.nprint("[+] PE entry point at 0x%x" % entry_point)
+                self.ql.modules.append((path, entry_point, pe))
+                return True
+            except UcError as e:
+                if e.errno == UC_ERR_MAP:
+                    IMAGE_BASE += 0x10000
+                    pe.relocate_image(IMAGE_BASE)
+                else:
+                    raise
+            except QlMemoryMappedError:
+                IMAGE_BASE += 0x10000
+                pe.relocate_image(IMAGE_BASE)
+        return False
+
+    def load(self):
+
+        """
+        initiate UC needs to be in loader, or else it will kill execve
+        Note: This is EFI, but for the sake of same with others OS
+        """
+        self.ql.uc = self.ql.arch.init_uc
+
+        if self.ql.archtype== QL_X8664:
+            self.QL_UEFI_STACK_ADDRESS = 0x7ffffffde000
+            self.QL_UEFI_STACK_SIZE = 0x40000
+            
+        elif self.ql.archtype== QL_X86:        
+            self.QL_UEFI_STACK_ADDRESS = 0xfffdd000
+            self.QL_UEFI_STACK_SIZE =0x21000 
+
+        if self.ql.stack_address == 0:
+            self.ql.stack_address = self.QL_UEFI_STACK_ADDRESS
+        if self.ql.stack_size == 0:
+            self.ql.stack_size = self.QL_UEFI_STACK_SIZE
+
+        if self.ql.path and not self.ql.shellcoder:
+            
+            # set stack pointer
+            self.ql.nprint("[+] Initiate stack address at 0x%x " % self.ql.stack_address)
+            self.ql.mem.map(self.ql.stack_address, self.ql.stack_size)
+
+            # Stack should not init at the very bottom. Will cause errors with Dlls
+            sp = self.ql.stack_address + self.ql.stack_size - 0x1000
+
+            if self.ql.archtype== QL_X86:
+                raise QlErrorArch("[!] Only 64 bit arch supported for now.")
+                # self.ql.register(UC_X86_REG_ESP, sp)
+                # self.ql.register(UC_X86_REG_EBP, sp)
+
+            elif self.ql.archtype== QL_X8664:
+                self.ql.register(UC_X86_REG_RSP, sp)
+                self.ql.register(UC_X86_REG_RBP, sp)
+
+            else:
+                raise QlErrorArch("[!] Unknown ql.arch")
+
+            if len(self.ql.argv) > 1:
+                for dependency in self.ql.argv[1:]:
+                    if not self.map_and_load(dependency):
+                        raise QlErrorFileType("Can't map dependency")
+
+            # Load main module
+            self.map_and_load(self.ql.path)
+
+            self.ql.nprint("[+] Done with loading %s" % self.ql.path)
+            self.filepath = b"D:\\" + bytes(self.ql.path.replace("/", "\\"), "utf-8")
+
+        elif self.ql.shellcoder:
+            # setup stack memory
+            self.ql.mem.map(self.ql.stack_address, self.ql.stack_size)
+            if self.ql.archtype== QL_X86:
+                self.ql.register(UC_X86_REG_ESP, self.ql.stack_address + 0x3000)
```

Comment:
Hi,

we also updated this.

new register will be 

ql.reg.esp for read and ql.reg.esp = 0x11 for write

you can also use 

ql.reg.read() and ql.reg.write

sorry for the mass we've created.